### PR TITLE
Add RAML and integration tests

### LINKED ISSUE #390: Create prototype API spec and testing process
Remaining tasks:

- [x] Get as many auto-generated integration tests passing as possible without large investment
- [x] Test documentation hosting strategy
- [x] Outline new endpoint design and mocking strategy

### PR DESCRIPTION
**Pull request by coltonlw**
Fixes #390

Adds raml/ directory with RAML api spec
Schemas moved from api/schemas to raml/schemas
Add Abao to Travis build for automated integration testing based on API spec
Add Abao test hooks to enable some very basic workflow testing

Ref: #404 , keep making progress on integration tests

Ref: #63 , This issue is not fixed yet but with RAML implemented, it's just a matter of busy work to fill out the rest of the spec.

The only changes to the code in this PR are moving schemas/ directory and adding new-user.json to the list of input schemas (both in config.py). Anything we can do to review / verify those changes is welcome.

### CODE REVIEW

**Comment by coltonlw** [comment]
@nagem Thanks for helping me with rebase!!

**Comment by coltonlw** [comment]
This was closed yesterday when a test began failing intermittently.  Huge thanks to @ryansanford for helping me track that down, the issue is documented here:  https://github.com/BigstickCarpet/json-schema-ref-parser/issues/26

I ran through 1000 iterations of the tests to make sure there are no more intermittent failures and the workaround is successful

File: raml/resources/download.raml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      type: string
+      required: true
+      description: ID of the download ticket
+  responses:
+    200:
+      description: The requested tarball download as a binary stream
+      body:
+        application/octet-stream:
+    400:
+      description: Ticket not for this source IP
```
Comment (nagem) [inline raml/resources/download.raml]: Does RAML spec allow for multiple instances of the same error code with different descriptions? For example, documenting an endpoint with 4** errors including `400: <some param> is required` as well as `400: <some other param> is not a valid date`.

Comment (nagem) [inline raml/resources/download.raml]: Or would the standard be to list all error responses in the same description?

File: raml/resources/jobs.raml [hunk 2]
```diff
@@ -0,0 +1,187 @@
+description: Run Gears on distributed computing
+get:
+  description: List all jobs
+  responses:
+    200:
+      body:
+        application/json:
+          example: !include ../examples/job_list.json
+          schema: |
```
Comment (nagem) [inline raml/resources/jobs.raml]: I believe we would want to move this to a separate schema file, correct?

Comment (coltonlw) [inline raml/resources/jobs.raml]: Yeap, I'll do that

File: raml/resources/download.raml (hunk 1 above)
Comment (coltonlw) [inline raml/resources/download.raml]: We want to have a particular status code represent only one error condition wherever possible.  For instance, I choose to use 400 for un-parseable JSON, then I choose to use 422 to represent json that did not match the schema.  400 would be a valid status code for the second condition, but then the status code becomes ambiguous.  In cases where one status code must be used and could represent different situations, the response should contain a body with information describing the specific condition.  Thus documenting multiple descriptions for the same error code isn't something we would ever do.

Comment (coltonlw) [inline raml/resources/download.raml]: Granted if an existing implementation of an endpoint uses status codes in an ambiguous way, and we aren't changing the status code behavior at that time, documenting the different responses in the same description is the best thing

**Comment by coltonlw** [comment]
@nagem I added a commit because userhandler was not actually using user-new.json .  I also added input/user-update.json in place of input/user.json   That issue might have contributed to the confusion about user schemas.  Can you review  schemas/input/user-new.json and  schemas/input/user-update.json now that I've made this change?

**Comment by nagem** [comment]
An overview of the PR LGTM, nice work Colton!

**Comment by ryansanford** [comment]
@coltonlw 
I disabled the httpdomain portion of the sphinx doc building, and also the auto-building of the docs in the docker environment.

From my perspective there are a couple TODOs for this PR.
1. Remove the httpdomain related docstrings from the API. Search for `.. http:` to find them all. Replace with python-centric docstrings that describe the function.
2. Create a way to test the html display of raml changes before they are pushed to github. Maybe as simple as a script that pulls down the `gh-pages` branch and tweaks `raml_url` in the index.html?

**Comment by coltonlw** [comment]
1) - Will do
2) - Gunnar had me add a "branch" query param to the version 1 api console page on github pages, so a developer can see the docs built from their feature branch.  The docs can also be previewed with the anypoint api manager and the api-workbench atom add-on.    http://scitran.github.io/core/docs/api/v1/index.html?branch=add-raml-testing

**Comment by coltonlw** [comment]
@ryansanford Got the docstrings removed for endpoints that exist in RAML
@gsfr I will work on getting google docs converted to markdown and made publicly accessible

**Comment by coltonlw** [comment]
Keep in mind this pull request and #414 are incompatible, in that megan adds a schema in the old location and I move the schemas folder.  Travis will notice this but just be aware that whoever merges second needs to rebase and fix that before merging.

**Comment by coltonlw** [comment]
Closing while I fix up new abao tests for jobs RAML I added

**Comment by coltonlw** [comment]
@gsfr I was able to get CONTRIBUTING.md and TESTING.md added, as well as README.md updated with links to everything

**Comment by coltonlw** [comment]
Closing to rebase after analyses pr merge

**Comment by coltonlw** [comment]
Closing while I add a test for megan's changes to analysis upload

**Comment by coltonlw** [comment]
@nagem I got the test for uploading to /engine with "level=analysis" added to this branch

@ryansanford This is ready to merge pending any final review

**Pull request merged**